### ISSUE TITLE
bugprone-redundant-branch-condition incorrectly marks a condition as redundant

### ISSUE THREAD

**Issue by igor-akhmetov** [llvm/llvm-project#205685]
```c++
int f();
void g();

void h(bool b)
{
    if (b)
    {
        for (int i = 0; i < 3; ++i)
        {
            switch (f())
            {
            case 4:
            case 5:
                if (b)  // Redundant condition 'b' [bugprone-redundant-branch-condition]
                {
                    b = false;
                    g();
                }
                break;
            }
        }
    }
}
```
https://gcc.godbolt.org/z/TqxoT9PYz

**Comment by dingcyrus** [comment]
The outer if (b) (line 6) ensures that the loop body is entered only when b is true. Inside the loop, no statement modifies b before reaching the inner if (b) (line 14): the return value of f() only affects the switch branch and does not change b; and the assignment b = false within the case 4 and case 5 branches is placed inside the condition block of that inner if, so it is executed only after b is already checked to be true. Therefore, when execution reaches the if (b) at line 14, b is necessarily true, making the condition always true and thus redundant.

**Comment by chrchr-github** [comment]
Imagine `f()` returns `5, 5, 5`. With the condition `if (b)`, `g()` is called once, without it, three times. So `if (b)` is not redundant.

**Comment by NagyDonat** [comment]
As @chrchr-github confirmed it, this is indeed a true positive.

The implementation in `RedundantBranchConditionCheck.cpp` checks whether `b` (i.e. the variable which is the condition of both nested `if` statements) is changed between the two `if` conditions, but interprets "between" in a syntactic sense (as the source code is written) instead of considering the execution order. (In the code example, the modification of `b` is syntactically not between the two conditions, but it _is_ between the evaluation of the outer condition and the second evaluation of the inner condition.)

To fix this, the check should notice the presence of loop "layers" between the two `if` conditions and if there are loops, it should look for mutating statements between the outer `if` condition and the end of the outermost loop around the inner if that is contained in the outer `if` block.

E.g. in the following code, the check should look for modifications of `b` between `BEGIN` and `END`
```c++
void f(int b) {
  while (a()) {
    if (b) { // BEGIN
      if (unrelated()) {
        for (int x : get_list()) {
          do {
            if (b) { /*body*/ }
          } while (bar);
          // ... some code
        } // END
        // ...
      }
      // ...
    }
  }
}
```

Unfortunately I don't have time to implement a fix for this bug, but I'm happy to help with reviews if somebody decides to work on this.

**Comment by mamadou-wane** [comment]
working on this
